lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
-		- `parent`: The Roblox object to contain the contained instances
-		- `key`: The Name to give the Roblox instance that gets created
-		- `context`: Used to pass Roact context values down the tree
+	assert(Type.is(node, Type.Node))
 
-	The structure created by this method is important to the functionality of
-	the reconciliation methods; they depend on this structure being well-formed.
-]]
-function Reconciler._mountInternal(element, parent, key, context)
-	if isPrimitiveElement(element) then
-		-- Primitive elements are backed directly by Roblox Instances.
-
-		local rbx = Instance.new(element.component)
-
-		-- Update Roblox properties
-		for key, value in pairs(element.props) do
-			Reconciler._setRbxProp(rbx, key, value, element)
-		end
+	return function(tree)
+		local nodesToVisit = {node}
+		local visitIndex = 1
+		local nodesToDestroy = {}
 
-		-- Create children!
-		local children = {}
+		while true do
+			local visitingNode = nodesToVisit[visitIndex]
 
-		if element.props[Core.Children] then
-			for key, childElement in pairs(element.props[Core.Children]) do
-				local childInstance = Reconciler._mountInternal(childElement, rbx, key, context)
-
-				children[key] = childInstance
+			if visitingNode == nil then
+				break
 			end
-		end
-
-		-- This name can be passed through multiple components.
-		-- What's important is the final Roblox Instance receives the name
-		-- It's solely for debugging purposes; Roact doesn't use it.
-		if key then
-			rbx.Name = key
-		end
 
-		rbx.Parent = parent
-
-		-- Attach ref values, since the instance is initialized now.
-		applyRef(element.props[Core.Ref], rbx)
+			for _, childNode in pairs(node.children) do
+				table.insert(nodesToVisit, childNode)
+			end
 
-		return {
-			[isInstanceHandle] = true,
-			_key = key,
-			_parent = parent,
-			_element = element,
-			_context = context,
-			_children = children,
-			_rbx = rbx,
-		}
-	elseif isFunctionalElement(element) then
-		-- Functional elements contain 0 or 1 children.
-
-		local instanceHandle = {
-			[isInstanceHandle] = true,
-			_key = key,
-			_parent = parent,
-			_element = element,
-			_context = context,
-		}
+			table.insert(nodesToDestroy, visitingNode)
 
-		local vdom = element.component(element.props)
-		if vdom then
-			instanceHandle._child = Reconciler._mountInternal(vdom, parent, key, context)
+			visitIndex = visitIndex + 1
 		end
 
-		return instanceHandle
-	elseif isStatefulElement(element) then
-		-- Stateful elements have 0 or 1 children, and also have a backing
-		-- instance that can keep state.
-
-		-- We separate the instance's implementation from our handle to it.
-		local instanceHandle = {
-			[isInstanceHandle] = true,
-			_key = key,
-			_parent = parent,
-			_element = element,
-			_child = nil,
-		}
-
-		local instance = element.component._new(element.props, context)
+		-- Destroy from back-to-front in order to destroy the nodes deepest in
+		-- the tree first.
+		for i = #nodesToDestroy, 1, -1 do
```

Comment:
This rewrite is trying to avoid recursion completely, which gives us more flexibility with deeper trees. I think it's also a little bit easier to reason about in some cases.